VERIFY on concurent write and table rename

E   VERIFY failed (2025-08-20T15:19:38.688490Z): tablet_id=72075186224038010;self_id=[5:7540687620353763600:2392];ev=NActors::IEventHandle;tablet_id=72075186224038010;event=TEvWrite;verification=internalPathId;fline=columnshard__write.cpp:440;
E   ydb/library/actors/core/log.cpp:855
E   ~TVerifyFormattedRecordWriter(): requirement false failed
E   0. /-S/util/system/yassert.cpp:83: NPrivate::InternalPanicImpl(int, char const*, char const*, int, int, int, TBasicStringBuf<char, std::__y1::char_traits<char>>, char const*, unsigned long) @ 0xA6D4165
E   1. /-S/util/system/yassert.cpp:55: NPrivate::Panic(NPrivate::TStaticBuf const&, int, char const*, char const*, char const*, ...) @ 0xA6CCFAC
E   2. /-S/ydb/library/actors/core/log.cpp:855: NActors::TVerifyFormattedRecordWriter::~TVerifyFormattedRecordWriter() @ 0xB56E7A3
E   3. /-S/ydb/core/tx/columnshard/columnshard__write.cpp:440: NKikimr::NColumnShard::TColumnShard::Handle(TAutoPtr<NActors::TEventHandle<NKikimr::NEvents::TDataEvents::TEvWrite>, TDelete>&, NActors::TActorContext const&) @ 0x15032A5D
E   4. /-S/ydb/core/tx/columnshard/columnshard_impl.h:442: NKikimr::NColumnShard::TColumnShard::StateWork(TAutoPtr<NActors::IEventHandle, TDelete>&) @ 0x14FB249E
E   5. /-S/ydb/library/actors/core/actor.cpp:406: NActors::IActor::Receive(TAutoPtr<NActors::IEventHandle, TDelete>&) @ 0xB526077
E   6. /-S/ydb/library/actors/core/executor_thread.cpp:268: NActors::TExecutorThread::Execute(NActors::TMailbox*, bool) @ 0xB560013
E   7. /-S/ydb/library/actors/core/executor_thread.cpp:458: NActors::TExecutorThread::ProcessExecutorPool()::$_0::operator()(NActors::TMailbox*, bool) const @ 0xB563B21
E   8. /-S/ydb/library/actors/core/executor_thread.cpp:510: NActors::TExecutorThread::ProcessExecutorPool() @ 0xB563770
E   9. /-S/ydb/library/actors/core/executor_thread.cpp:536: NActors::TExecutorThread::ThreadProc() @ 0xB56431A
E   10. /-S/util/system/thread.cpp:245: (anonymous namespace)::TPosixThread::ThreadProxy(void*) @ 0xA6D4FBC
E   11. ./nptl/pthread_create.c:444: start_thread @ 0x7FC54C046B59
E   12. ../sysdeps/unix/sysv/linux/x86_64/clone3.S:78: clone3 @ 0x7FC54C0D75FB

--- a/ydb/core/tx/columnshard/columnshard__write.cpp
+++ b/ydb/core/tx/columnshard/columnshard__write.cpp
@@ -437,7 +437,11 @@ void TColumnShard::Handle(NEvents::TDataEvents::TEvWrite::TPtr& ev, const TActor
 
     const auto schemeShardLocalPathId = TSchemeShardLocalPathId::FromProto(operation.GetTableId());
     const auto& internalPathId = TablesManager.ResolveInternalPathId(schemeShardLocalPathId, false);
-    AFL_VERIFY(internalPathId);
+    if (!internalPathId) {
+        LWPROBE(EvWrite, TabletID(), source.ToString(), cookie, record.GetTxId(), writeTimeout.value_or(TDuration::Max()), 0, "", false, operation.GetIsBulk(), ToString(NKikimrDataEvents::TEvWriteResult::STATUS_BAD_REQUEST), "unknown table");
+        sendError("unknown table", NKikimrDataEvents::TEvWriteResult::STATUS_BAD_REQUEST);
+        return;
+    }
     const auto& pathId = TUnifiedPathId::BuildValid(*internalPathId, schemeShardLocalPathId);
     if (!TablesManager.IsReadyForStartWrite(*internalPathId, false)) {
         LWPROBE(EvWrite, TabletID(), source.ToString(), cookie, record.GetTxId(), writeTimeout.value_or(TDuration::Max()), 0, "", false, operation.GetIsBulk(), ToString(NKikimrDataEvents::TEvWriteResult::STATUS_INTERNAL_ERROR), "unknown schema version");

--- a/ydb/core/tx/columnshard/columnshard__write.cpp
+++ b/ydb/core/tx/columnshard/columnshard__write.cpp
@@ -439,6 +439,7 @@ void TColumnShard::Handle(NEvents::TDataEvents::TEvWrite::TPtr& ev, const TActor
     const auto& internalPathId = TablesManager.ResolveInternalPathId(schemeShardLocalPathId, false);
     if (!internalPathId) {
         LWPROBE(EvWrite, TabletID(), source.ToString(), cookie, record.GetTxId(), writeTimeout.value_or(TDuration::Max()), 0, "", false, operation.GetIsBulk(), ToString(NKikimrDataEvents::TEvWriteResult::STATUS_BAD_REQUEST), "unknown table");
+        AFL_WARN(NKikimrServices::TX_COLUMNSHARD_WRITE)("event", "unknown_table")("path_id", schemeShardLocalPathId);
         sendError("unknown table", NKikimrDataEvents::TEvWriteResult::STATUS_BAD_REQUEST);
         return;
     }

--- a/ydb/core/tx/columnshard/ut_schema/ut_columnshard_move_table.cpp
+++ b/ydb/core/tx/columnshard/ut_schema/ut_columnshard_move_table.cpp
@@ -39,8 +39,8 @@ Y_UNIT_TEST_SUITE(MoveTable) {
         TActorId sender = runtime.AllocateEdgeActor();
 
         const ui64 srcPathId = 1;
-        TestTableDescription testTabe{};
-        auto planStep = PrepareTablet(runtime, srcPathId, testTabe.Schema);
+        TestTableDescription testTable{};
+        auto planStep = PrepareTablet(runtime, srcPathId, testTable.Schema);
 
         ui64 txId = 10;
         const ui64 dstPathId = 2;
@@ -55,31 +55,31 @@ Y_UNIT_TEST_SUITE(MoveTable) {
         TActorId sender = runtime.AllocateEdgeActor();
 
         const ui64 srcPathId = 1;
-        TestTableDescription testTabe{};
-        auto planStep = PrepareTablet(runtime, srcPathId, testTabe.Schema);
+        TestTableDescription testTable{};
+        auto planStep = PrepareTablet(runtime, srcPathId, testTable.Schema);
 
         ui64 txId = 10;
         int writeId = 10;
         std::vector<ui64> writeIds;
-        const bool ok = WriteData(runtime, sender, writeId++, srcPathId, MakeTestBlob({0, 100}, testTabe.Schema), testTabe.Schema, true, &writeIds);
+        const bool ok = WriteData(runtime, sender, writeId++, srcPathId, MakeTestBlob({0, 100}, testTable.Schema), testTable.Schema, true, &writeIds);
         UNIT_ASSERT(ok);
         const ui64 dstPathId = 2;
         planStep = ProposeSchemaTx(runtime, sender, TTestSchema::MoveTableTxBody(srcPathId, dstPathId, 1), ++txId);
         {
             const bool ok =
-                WriteData(runtime, sender, writeId++, srcPathId, MakeTestBlob({ 100, 200 }, testTabe.Schema), testTabe.Schema, true, &writeIds);
+                WriteData(runtime, sender, writeId++, srcPathId, MakeTestBlob({ 100, 200 }, testTable.Schema), testTable.Schema, true, &writeIds);
             UNIT_ASSERT(!ok);
         }
         PlanSchemaTx(runtime, sender, {planStep, txId});
         {
             TShardReader reader(runtime, TTestTxConfig::TxTablet0, srcPathId, NOlap::TSnapshot(planStep, txId));
-            reader.SetReplyColumnIds(TTestSchema::ExtractIds(testTabe.Schema));
+            reader.SetReplyColumnIds(TTestSchema::ExtractIds(testTable.Schema));
             auto rb = reader.ReadAll();
             UNIT_ASSERT(!rb);
         }
         {
             TShardReader reader(runtime, TTestTxConfig::TxTablet0, dstPathId, NOlap::TSnapshot(planStep, txId));
-            reader.SetReplyColumnIds(TTestSchema::ExtractIds(testTabe.Schema));
+            reader.SetReplyColumnIds(TTestSchema::ExtractIds(testTable.Schema));
             auto rb = reader.ReadAll();
             UNIT_ASSERT(!rb);
         }
@@ -93,14 +93,14 @@ Y_UNIT_TEST_SUITE(MoveTable) {
         TActorId sender = runtime.AllocateEdgeActor();
 
         const ui64 srcPathId = 1;
-        TestTableDescription testTabe{};
-        auto planStep = PrepareTablet(runtime, srcPathId, testTabe.Schema);
+        TestTableDescription testTable{};
+        auto planStep = PrepareTablet(runtime, srcPathId, testTable.Schema);
 
         ui64 txId = 10;
         int writeId = 10;
         std::vector<ui64> writeIds;
         const bool ok =
-            WriteData(runtime, sender, writeId++, srcPathId, MakeTestBlob({ 0, 100 }, testTabe.Schema), testTabe.Schema, true, &writeIds);
+            WriteData(runtime, sender, writeId++, srcPathId, MakeTestBlob({ 0, 100 }, testTable.Schema), testTable.Schema, true, &writeIds);
         UNIT_ASSERT(ok);
         planStep = ProposeCommit(runtime, sender, ++txId, writeIds);
         PlanCommit(runtime, sender, planStep, txId);
@@ -119,7 +119,7 @@ Y_UNIT_TEST_SUITE(MoveTable) {
 
         {
             TShardReader reader(runtime, TTestTxConfig::TxTablet0, dstPathId, NOlap::TSnapshot(planStep, txId));
-            reader.SetReplyColumnIds(TTestSchema::ExtractIds(testTabe.Schema));
+            reader.SetReplyColumnIds(TTestSchema::ExtractIds(testTable.Schema));
             auto rb = reader.ReadAll();
             UNIT_ASSERT(rb);
             UNIT_ASSERT_EQUAL(rb->num_rows(), 100);
@@ -127,7 +127,7 @@ Y_UNIT_TEST_SUITE(MoveTable) {
 
         {
             TShardReader reader(runtime, TTestTxConfig::TxTablet0, srcPathId, NOlap::TSnapshot(planStep, txId));
-            reader.SetReplyColumnIds(TTestSchema::ExtractIds(testTabe.Schema));
+            reader.SetReplyColumnIds(TTestSchema::ExtractIds(testTable.Schema));
             auto rb = reader.ReadAll();
             UNIT_ASSERT(!rb);
         }
@@ -141,8 +141,8 @@ Y_UNIT_TEST_SUITE(MoveTable) {
         TActorId sender = runtime.AllocateEdgeActor();
 
         const ui64 srcPathId = 1;
-        TestTableDescription testTabe{};
-        const auto& planStep = PrepareTablet(runtime, srcPathId, testTabe.Schema);
+        TestTableDescription testTable{};
+        const auto& planStep = PrepareTablet(runtime, srcPathId, testTable.Schema);
         Y_UNUSED(planStep);
 
         const ui64 absentPathId = 111;
@@ -158,8 +158,8 @@ Y_UNIT_TEST_SUITE(MoveTable) {
         TActorId sender = runtime.AllocateEdgeActor();
 
         const ui64 srcPathId = 1;
-        TestTableDescription testTabe{};
-        const auto& planStep = PrepareTablet(runtime, srcPathId, testTabe.Schema);
+        TestTableDescription testTable{};
+        const auto& planStep = PrepareTablet(runtime, srcPathId, testTable.Schema);
         Y_UNUSED(planStep);
         ui64 txId = 10;
         ProposeSchemaTxFail(runtime, sender, TTestSchema::MoveTableTxBody(srcPathId, srcPathId, 1), ++txId);

--- a/ydb/core/tx/columnshard/ut_schema/ut_columnshard_move_table.cpp
+++ b/ydb/core/tx/columnshard/ut_schema/ut_columnshard_move_table.cpp
@@ -65,7 +65,25 @@ Y_UNIT_TEST_SUITE(MoveTable) {
         UNIT_ASSERT(ok);
         const ui64 dstPathId = 2;
         planStep = ProposeSchemaTx(runtime, sender, TTestSchema::MoveTableTxBody(srcPathId, dstPathId, 1), ++txId);
+        {
+            const bool ok =
+                WriteData(runtime, sender, writeId++, srcPathId, MakeTestBlob({ 100, 200 }, testTabe.Schema), testTabe.Schema, true, &writeIds);
+            UNIT_ASSERT(!ok);
+        }
         PlanSchemaTx(runtime, sender, {planStep, txId});
+        {
+            TShardReader reader(runtime, TTestTxConfig::TxTablet0, srcPathId, NOlap::TSnapshot(planStep, txId));
+            reader.SetReplyColumnIds(TTestSchema::ExtractIds(testTabe.Schema));
+            auto rb = reader.ReadAll();
+            UNIT_ASSERT(!rb);
+        }
+        {
+            TShardReader reader(runtime, TTestTxConfig::TxTablet0, dstPathId, NOlap::TSnapshot(planStep, txId));
+            reader.SetReplyColumnIds(TTestSchema::ExtractIds(testTabe.Schema));
+            auto rb = reader.ReadAll();
+            UNIT_ASSERT(!rb);
+        }
+
     }
 
     Y_UNIT_TEST_DUO(WithData, Reboot) {

